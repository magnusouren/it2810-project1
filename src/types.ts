export type CategoryType =
  | 'Ordinary drink'
  | 'Cocktail'
  | 'Shake'
  | 'Other / Unknown'
  | 'Cocoa'
  | 'Shot'
  | 'Coffee / Tea'
  | 'Homemade Liqueur'
  | 'Punch / Party Drink'
  | 'Beer'
  | 'Soft Drink';

export type Alcoholic = 'Alcoholic' | 'Non alcoholic' | 'Optional alcohol';

export interface Ingredient {
  ingredient: string;
  measure: string;
}

export interface Drink {
  drinkId: string;
  strDrink: string;
  ingredients: Array<Ingredient>;
  strGlass: string;
  strInstructions: string;
  strDrinkThumb: string;
  strCategory: string;
  strAlcoholic: Alcoholic;
}
