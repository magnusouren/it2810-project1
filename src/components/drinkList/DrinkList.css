.drink-div {
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  width: 60%;
  gap: 1rem;
}

h2 {
  width: 60%;
  margin: 0 auto 1rem;
  padding-bottom: 0.5rem;
  border-bottom: solid 2px #f2c2a6;
}

@media screen and (width <= 549px) {
  .drink-div {
    width: 90%;
    display: block;
  }
}
