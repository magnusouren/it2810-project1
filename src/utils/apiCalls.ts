import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
