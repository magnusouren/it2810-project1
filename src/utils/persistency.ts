import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
