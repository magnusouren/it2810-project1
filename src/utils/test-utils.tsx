import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
