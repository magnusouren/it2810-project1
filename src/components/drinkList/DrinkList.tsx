import './DrinkList.css';

import { FC } from 'react';

import { SimpleDrink } from '../../types';

interface DrinkListProps {
  drinks: SimpleDrink[] | undefined;
}

export const DrinkList: FC<DrinkListProps> = ({ drinks }) => {
  return (
    <div>
      {drinks && (
        <ul className='drink-list'>
          {drinks.map((drink) => (
            <li key={drink.drinkId}>{drink.strDrink}</li>
          ))}
        </ul>
      )}
    </div>
  );
};
