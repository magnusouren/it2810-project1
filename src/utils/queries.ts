import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
