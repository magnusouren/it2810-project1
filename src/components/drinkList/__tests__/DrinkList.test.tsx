import { SimpleDrinkType } from '../../../types';
import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { DrinkList } from '../DrinkList';

const simpleDrinks: SimpleDrinkType[] = [
  {
    strDrink: 'Margarita',
    strDrinkThumb: 'https://www.thecocktaildb.com/images/media/drink/5noda61589575158.jpg',
    idDrink: '11007',
  },
  {
    strDrink: 'Blue Margarita',
    strDrinkThumb: 'https://www.thecocktaildb.com/images/media/drink/bry4qh1582751040.jpg',
    idDrink: '178332',
  },
  {
    strDrink: "Tommy's Margarita",
    strDrinkThumb: 'https://www.thecocktaildb.com/images/media/drink/loezxn1504373874.jpg',
    idDrink: '17216',
  },
];

describe('DrinkList', () => {
  it('Should match snapshot', () => {
    const { container } = renderWithRouterAndQueryClient(<DrinkList drinks={simpleDrinks} />);
    expect(container).toMatchSnapshot();
  });

  it('Should render correct number of drinks', () => {
    renderWithRouterAndQueryClient(<DrinkList drinks={simpleDrinks} />);
    expect(document.getElementsByClassName('drink-div')[0].childElementCount).toBe(3);
  });

  it('Should render correct drinks', () => {
    renderWithRouterAndQueryClient(<DrinkList drinks={simpleDrinks} />);
    expect(document.getElementsByClassName('drink-div')[0].children[0].textContent).toBe('Margarita');
    expect(document.getElementsByClassName('drink-div')[0].children[1].textContent).toBe('Blue Margarita');
    expect(document.getElementsByClassName('drink-div')[0].children[2].textContent).toBe("Tommy's Margarita");
  });
});
