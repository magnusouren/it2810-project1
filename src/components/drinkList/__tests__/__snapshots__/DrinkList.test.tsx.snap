// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrinkList > Should match snapshot 1`] = `
<div>
  <div
    class="drink-div"
  >
    <a
      class="simple-drink"
      href="/drink/11007"
    >
      <h3>
        Margarita
      </h3>
      <img
        alt="Drink image"
        class="drink-image"
        src="https://www.thecocktaildb.com/images/media/drink/5noda61589575158.jpg"
      />
    </a>
    <a
      class="simple-drink"
      href="/drink/178332"
    >
      <h3>
        Blue Margarita
      </h3>
      <img
        alt="Drink image"
        class="drink-image"
        src="https://www.thecocktaildb.com/images/media/drink/bry4qh1582751040.jpg"
      />
    </a>
    <a
      class="simple-drink"
      href="/drink/17216"
    >
      <h3>
        Tommy's Margarita
      </h3>
      <img
        alt="Drink image"
        class="drink-image"
        src="https://www.thecocktaildb.com/images/media/drink/loezxn1504373874.jpg"
      />
    </a>
  </div>
</div>
`;
